### ISSUE TITLE
Introduce optional delay between changing resources in a loop (f.e. nodes in a cluster)

### ISSUE THREAD

**Issue by gdubicki** [opentofu/opentofu#550]
### OpenTofu Version

```shell
Any
```

### Use Cases

Often the resources managed in a loop (e.g. `for_each`) are a cluster that provides a single service.

F.e. a loop of nodes that are Redis cluster.

When you want to apply a change that forces a restart of a resource or its recreation, the service is often resilient to the unavailability of some number of nodes at a time.

However, when you do such a change with the current version of opentofu, it [does such changes by default for up to 10 nodes simultaneously](https://opentofu.org/docs/internals/graph#walking-the-graph).

### Attempted Solutions

You can change the `-parallelism=n` value, but it applies to the whole run instead of just a single group of resources.

It also does not allow to introduction of a delay that may be needed for a given cluster to get healthy after a single node restart or recreation.

### Proposal

Let's introduce an optional parameter for the `for_each` expression that will introduce a delay before each of the resources created with it is changed in a disrupting manner (restart or recreation).

**Comment by Hefeweizen** [comment]
This an interesting request.  Would you imagine the delay only on create/update actions, or would it need to be honored on destroy as well?

I'm also curious, have you ever experimented with `-target` and `-parallelsm=n` at the same time?  Does it provide the behaviour you're looking for?

**Comment by Magnitus-** [comment]
> Often the resources managed in a loop (e.g. `for_each`) are a cluster that provides a single service.
> 
> F.e. a loop of nodes that are Redis cluster.
> 
> When you want to apply a change that forces a restart of a resource or its recreation, the service is often resilient to the unavailability of some number of nodes at a time.
> 
> However, when you do such a change with the current version of opentofu, it [does such changes by default for up to 10 nodes simultaneously](https://opentofu.org/docs/internals/graph#walking-the-graph).

We are tackling similar problems, I've thought a fair amount about it and I think short of having a tool that understands how to talk to the distributed service you are managing, I wouldn't recommend just having a delay in between re-creations and assume nodes in a cluster will re-create and rejoin properly (especially if production data and availability are at stake).

Unless opentf greatly expands its functionality to include observability resources (having some kind of linguistic specification that express a need like "before you recreate the next etcd node, check that the cluster has 3 nodes and that none of them is a learner" which would definitely rely on third-party plugin architecture much like providers do), I think it is best to perform such maintenance across several apply commands.

That being said, there might be other sensible use cases for delays, like mitigating an api's rate limitation.

**Comment by gdubicki** [comment]
> This an interesting request. Would you imagine the delay only on create/update actions, or would it need to be honored on destroy as well?

That could be configurable, but in my use cases I only needed it for the updates.

> I'm also curious, have you ever experimented with `-target` and `-parallelsm=n` at the same time? Does it provide the behaviour you're looking for?

Yes, but we have used Terraform Cloud / Enterprise where you could not do both of these. And just setting `-parallelism=n` made the whole run super slow.

**Comment by gdubicki** [comment]
@Magnitus- : I agree that this would not comprehensively cover all the clusters' use cases.

But it would do it for a simpler use case, which I also had, where the nodes are not really a cluster but just a set of stateless app nodes for some service. The delay here would just ensure that there would be no downtime during the updates that require node restart.

**Comment by eranelbaz** [comment]
What about using `time_sleep` resource?

**Comment by Hefeweizen** [comment]
TIL `time_sleep`!  Thanks for mentioning it.

I wanted to experiment with this potential solution, but I'm stuck:
```
$ cat main.tf 
locals {
  targets = toset(["one", "two", "three"])
}

# This resource will destroy (potentially immediately) after null_resource.next
resource "null_resource" "previous" {
  for_each = local.targets
}

resource "time_sleep" "wait_30_seconds" {
  for_each = local.targets

  depends_on = [null_resource.previous[each.key]]

  create_duration = "30s"
}

# This resource will create (at least) 30 seconds after null_resource.previous
resource "null_resource" "next" {
  depends_on = [time_sleep.wait_30_seconds]
}
$ opentf plan -out=plan.out 
╷
│ Error: Invalid expression
│ 
│   on main.tf line 13, in resource "time_sleep" "wait_30_seconds":
│   13:   depends_on = [null_resource.previous[each.key]]
│ 
│ A single static variable reference is required: only attribute access and indexing with constant keys. No
│ calculations, function calls, template expressions, etc are allowed here.
╵
```
Did I get the syntax wrong, or is this a limitation in `depends_on`?

Even if this does work, I'm not sure it'll meet the need.  This creates a delay on each individual resource, but OP wants a delay between successive resources.  In order to implement successive resources, we'd need something like `each.key-1`

**Comment by eranelbaz** [comment]
I tried to use `provisioner "local-exec"`

```
$ cat main.tf
locals {
  seconds = toset(["1", "2", "3"])
}

# This resource will destroy (potentially immediately) after null_resource.next
resource "null_resource" "previous" {
  provisioner "local-exec" {
    command = "sh sleep.sh ${each.key} ${tonumber(each.key)}"
  }

  for_each = local.seconds
}

$ cat sleep.sh
#!/bin/bash
name=$1
sleep_seconds=$2

for ((i=1; i<=$sleep_seconds; i++)); do
  echo "[$name] Waiting for $i second(s)..."
  sleep 1
done

echo "[$name] Done waiting for $sleep_seconds seconds!"

$ tofu apply

null_resource.previous["3"]: Creating...
null_resource.previous["2"]: Creating...
null_resource.previous["1"]: Creating...
null_resource.previous["2"]: Provisioning with 'local-exec'...
null_resource.previous["1"]: Provisioning with 'local-exec'...
null_resource.previous["3"]: Provisioning with 'local-exec'...
null_resource.previous["2"] (local-exec): Executing: ["/bin/sh" "-c" "sh sleep.sh 2 2"]
null_resource.previous["3"] (local-exec): Executing: ["/bin/sh" "-c" "sh sleep.sh 3 3"]
null_resource.previous["1"] (local-exec): Executing: ["/bin/sh" "-c" "sh sleep.sh 1 1"]
null_resource.previous["3"] (local-exec): [3] Waiting for 1 second(s)...
null_resource.previous["1"] (local-exec): [1] Waiting for 1 second(s)...
null_resource.previous["2"] (local-exec): [2] Waiting for 1 second(s)...
null_resource.previous["3"] (local-exec): [3] Waiting for 2 second(s)...
null_resource.previous["1"] (local-exec): [1] Done waiting for 1 seconds!
null_resource.previous["2"] (local-exec): [2] Waiting for 2 second(s)...
null_resource.previous["1"]: Creation complete after 1s [id=9220235750353477881]
null_resource.previous["3"] (local-exec): [3] Waiting for 3 second(s)...
null_resource.previous["2"] (local-exec): [2] Done waiting for 2 seconds!
null_resource.previous["2"]: Creation complete after 2s [id=3230266246428695525]
null_resource.previous["3"] (local-exec): [3] Done waiting for 3 seconds!
null_resource.previous["3"]: Creation complete after 3s [id=8819875784296754600]
```

**Comment by eranelbaz** [comment]
@Hefeweizen I think the option with `time_sleep` is to export that to module and iterate over it

**Comment by Hefeweizen** [comment]
I'm not sure I see how wrapping it in a module helps.  Bundling the resource and the sleep together can be done in a number of ways, it's the depends_on that I'm having trouble with.  How do we tell resource[2] to depend on resource[1], and resource[3] to depend on resource[2] (or with names in a for_each)?  If we bundled things in a module, we'd still have to chain the depends_on so that module[2] depends on module[1], module[3] on module[2], &c.

**Comment by eranelbaz** [comment]
Because by wrapping it you can add the time sleep for each specific resource

**Comment by eranelbaz** [comment]
@gdubicki Does the discussion above solves your issue?

**Comment by Hefeweizen** [comment]
In this discussion, we've talked about two mechanisms for introducing delay.  But the delay is only to serve a purpose:

> ... a delay that may be needed for a given cluster to get healthy after a single node restart or recreation.

The delay is requested so that the cluster operations can proceed serially.  The delay is requested so that one new resource at a time is created.  The delay is intended to let the cluster integrate the new component before proceeding to the next.

As I see it, the request is to enable having a pattern of: add one new resource, wait, add a second resource, wait, add a third, wait, and so on.

If we bundle `time_sleep` or a `local_exec` into the resource creation, we're creating X resources at a time.  We're telling the cluster "here's X new resources, all at once", and the cluster must integrate X resources, rather than achieving the intended effect.  The delay is important, but chaining the `depends_on` in some also seems to be important.

**Comment by eranelbaz** [comment]
@Hefeweizen if the request is to create it serially that's not the described issue
let's wait for @gdubicki feedback as we suggested multiple ideas but we need to know the direction

**Comment by Hefeweizen** [comment]
@eranelbaz Certainly!

**Comment by kolesaev** [comment]
We'd also like be able not to depends_on a full object, but an instance of a object as well, like
```
module "module_1" {
  for_each = var.module_1_vars_map
  config = each.value
}

module "module_2" {
  for_each = var.module_2_vars_map
  config = each.value

  depends_on = [
    module.module_1[each.value.module_1_key]
  ]
}

module "module_3" {
  for_each = var.module_3_vars_map
  config = each.value

  depends_on = [
    module.module_1[each.value.module_2_key]
  ]
}

...and so on
```
So when we remove one item from module_1 and related item from modules 2 and 3, we would not get that all module_2 and module_3 instances destroyed with creating new instances for left configs.

With this feature we would be able to use terraform without terragrunt for one-kind multi-objects support (but not k8s multi-cluster objects yet, as there we will need another difficult feature or an own preparing script)

**Comment by janosdebugs** [comment]
Hey folks, thank you for the lively discussion on this topic. I have ran into this myself several times, so let me ask a clarifying question: for_each in TF is actually executing in parallel. Would a construct similar to for_each which ran sequentially do the trick? Does anyone know workarounds to run for_eaches sequentially?

**Comment by lorengordon** [comment]
I'd be curious whether exposing retry mechanisms in a more general way might also do the trick. Something to the effect of, do X at a time, but if you encounter any of these types of failures (message, return code, etc), then try again. The aws sdk does this automatically for many types of errors (rate limits), but might be nice to have a more general feature in the language as an escape hatch for sdks/apis that don't have good error handling.

**Comment by janosdebugs** [comment]
From what I remember from implementing the oVirt provider, the retries were solely a provider-specific issue and are not exposed to the OpenTofu core. Providers have a lot of freedom to implement things in any way they want.

**Comment by lorengordon** [comment]
Yes they do, and some providers for some resources do have specific/enhanced retry handling. But I'm talking about a core feature. Similar to the various lifecycle options.

**Comment by janosdebugs** [comment]
That would need a separate issue and some careful consideration, not every API call is repeatable and could lead to some very expensive results if done incorrectly.

**Comment by lorengordon** [comment]
:shrug: take it or leave it. Just another idea that addresses the underlying issue at hand, which is the failure, not really the need for a delay or sequential execution of resource instances in a for_each loop.

**Comment by janosdebugs** [comment]
This issue sounds to me like it's more calling for a gradual rollout of a change which would lead to an outage, less like it's intended to handle resource creation failures.

**Comment by apparentlymart** [comment]
The following is just some casual thinking aloud about use-cases related to this request and ideas about how we could meet different variations of them. Some of it is quite far afield from the actual feature request here and I'm sharing that only for context and not meaning to suggest that this issue's scope ought to grow to cover _all of this_ :point_down: .

This is also not an official statement from the OpenTofu core team, just my own thoughts and ideas.

---

As the :ghost: spooky participant :ghost: implied further up, it's a little misleading to think of what `for_each` does as a "loop" because that tends to lead us to think about repeating an operation multiple times _sequentially_ as in a procedural programming language, whereas OpenTofu's `count` and `for_each` is more like declaring multiple independent objects that all happen to have a similar configuration generated using the same set of rules.

OpenTofu currently treats them all as independent of each other, which has two important consequences for the sake of what we're discussing here:

- Actions for them can happen concurrently, assuming there are enough concurrency slots available.

    The (unfortunately-named) `-parallelism` option is chooses the number of concurrency slots, defaulting to 10 in today's OpenTofu.

- If there aren't enough concurrency slots available for all of the instances to take their actions together, the order that they start is _unspecified_ and varies from one run to the next.

    The ordering is ultimately decided by the Go runtime's scheduler, since the internal implementation is literally to start one [goroutine](https://go.dev/doc/effective_go#goroutines) for each instance and have them race to acquire one concurrency slot each. Any instances that lose the race are blocked until another concurrency slot is released. The concurrency slots are managed by incrementing and decrementing a global counting [semaphore](https://en.wikipedia.org/wiki/Semaphore_(programming)).

---

I can certainly understand the various desires to exert more control over this, which people have expressed in a lot of different ways over the years, including but not limited to:

1. Choosing a smaller number of concurrency slots _for a particular resource_, because one specific underlying API cannot support as many concurrent requests as others. It's annoying to have to lower the global concurrency limit and slow _everything_ down just to work around one API.

    Architecturally today it's supposed to be a provider's job to deal with this, by knowing the concurrency or rate limits of each API the provider is wrapping and abstracting these details away so that although OpenTofu _thinks_ it's making 10 concurrent requests they are actually being handled sequentially inside the provider plugin. It's ideal for a provider to deal with this problem itself because then the provider can potentially use signals generated by the remote API, such as when a rate limit error includes information about when the rate limit will reset, to retry _as soon as possible_ rather than only after a fixed delay.

    However, in practice it's often hard to persuade provider developers to invest time in implementing this sort of thing, because (at least for the most commonly-used providers) it's an uphill struggle just to keep up with the constant stream of new features being added to each of the major cloud platforms and other services and so they need to prioritize carefully.

    As long as the problem is limited only to limiting concurrency and doesn't require _also_ controlling the order that the items are visited, I think there is a plausible answer to this one that isn't super hard to implement:

    ```hcl
    resource "whatever_example" "example" {
      for_each = ...

      lifecycle {
        instance_concurrency_limit = 1
      }
    }
    ```

    We could implement this by giving each resource where `instance_concurrency_limit` is set its own secondary counting semaphore, which each of the graph nodes representing the instances of this resource have a pointer to. The "execute" implementation for those graph nodes could then check whether the pointer to the secondary semaphore is non-nil and if so block until they have acquired the secondary semaphore.

    This would effectively reproduce the same model we currently have _globally_ but scoped to a particular resource. The per-resource instance concurrency limit would be separate from the global concurrency limit chosen by the `-parallelism` option, but the two would still interact because a graph node representing a resource instance won't even get a chance to acquire its secondary semaphore until the overall execution engine is able to acquire the global semaphore. Although the actual details would be a little more complicated than this in practice, it probably suffices to just imagine that the concurrency limit for a particular resource is _more-or-less_ the minimum of its own concurrency limit and the global concurrency limit.

2. Forcing a particular resource to have its instances visited one at a time _in an explicitly-chosen order_, because `for_each` is being used to describe a series of objects that buiild on their predecessors rather than a set of independent objects.

    This is just not supported _at all_ today, either in OpenTofu or in provider plugins. The design of the OpenTofu language is oriented around data-flow (dependencies between objects) rather than explicit control flow (as in procedural languages) and so our language doesn't really have good tools today for tightly controlling the order that operations should occur. A provider can't implement this itself because a provider (intentionally) cannot tell whether two concurrent requests are for two instances of the same resource or instances of two unrelated resources.

    While I would like to find some way to support this as a language feature, it's been challenging to find a syntax for it that is both intelligible and flexible enough to meet all of the use-cases that might lead to this request, particularly in shared modules where the required order is often chosen dynamically by the caller rather than something that can be hard-coded directly in the resource.

    There is also a further challenge: whereas in a procedural programming language all iterations of the loop are always performed, in OpenTofu it's quite possible that only a subset of the instances of a resource have a planned action in a given plan, and they might not all have the same action. In the _worst_ case we can have a mixture of "update" and "replace" actions where it's then often ambiguous what the correct sequence even _is_: just describing an order for the instances is no longer sufficient, because some of the instances have two actions to perform -- "replace" is really a compound action of destroy followed by create or the other way around -- and in real use-cases it's often very important what order these operations happen in.

    It's possible that _this particular situation_ is just a poor fit for OpenTofu's execution model, and that it would be better to use a different tool that uses an imperative programming style when you need such tight control over the order of operations. Perhaps, though, we could find a way to integrate with such a tool so that it's more ergonomic to mix OpenTofu-style data flow style and some other tool's imperative style, perhaps using OpenTofu-style planned actions only to decide whether to run the separate imperative tool in a particular round.

    [A very old, long-abandoned proposal in the predecessor project](https://github.com/hashicorp/terraform/issues/6810) had me thinking about such things, but there were still plenty of gaps to deal with. The broad idea behind that was to be able to express ideas like "execute this sequential process during the apply phase whenever some condition is true about other parts of the plan", but the finer details of that are kinda hard to define in a way that's general enough to meet many of the use-case variations.

3. Declaring arbitrary dependencies between the instances of a resource, while still potentially allowing some to be visited concurrently in an unspecified order as today, _and_ to propagate data from one instance to another. For example, this request often arises when OpenTofu is being used to describe arbitrary tree structures like with [`aws_api_gateway_resource`](https://library.tf/providers/hashicorp/aws/latest/docs/resources/api_gateway_resource) where `parent_id` wants to be populated by a reference to another resource (in the API gateway sense, rather than the OpenTofu sense).

    Interestingly, this capability would _mostly_ functionally subsume the previous one: if you can declare dependencies between instances then you can declare a chain of instances that each depends on the one before it and thus force a particular sequence. It's debatable how intuitive/ergonomic it is to specify a sequential process as a dependency graph, though, and this still doesn't directly solve the question of how to deal with mixtures of different actions for different instances in the same plan.

    The big design challenge with _this_ request is that in real cases the dependencies between the instances need to be decided dynamically based on data collected at runtime. In the API gateway resource example, the `id` of a resource is not decided until it's created during the apply phase, so a module trying to describe a tree of resources would need to use some _other_ kind of identifier, probably specified statically in the configuration, to describe the dependency relationships to OpenTofu.

    This particular idea has a relatively-clear proposal for _surface-level syntax_, but is very challenging from an implementation perspective.

    For example, perhaps we could allow an expression in a `resource` block to refer back to the same resource (which is currently forbidden) _if and only if_ the reference is followed by indexing brackets containing an expression that produces a valid instance key for the resource:

    ```hcl
    variable "api_resources" {
      type = map(object({
        path_part  = string
        parent_key = optional(string)
      }))
    }

    resource "aws_api_gateway_resource" "example" {
      for_each = var.api_resources

      rest_api_id = aws_api_gateway_rest_api.main.id

      path_part = each.value.path_part
      parent_id = (
        each.value.parent_key != null ?
        aws_api_gateway_resource.example[each.value.parent_key] :
        aws_api_gateway_rest_api.main.root_resource_id
      )
    }
    ```

    I _think_ it's pretty intuitive what the above is intended to mean: if a particular resource instance has a non-null value for `each.value.parent_key` then it depends on whichever other instance has the key specified in that attribute, so a parent will always be created before its children and thus the parent id will be known by the time the children are being created.

    Actually making OpenTofu detect and act on that information is _far_ easier said than done, though: OpenTofu currently builds a dependency graph using just static analysis (because the dependency graph is how we ensure that expressions in the configuration are evaluated in the correct order) but in this case we can't decide the dependency graph without evaluating `each.value.parent_key` (more generally: _whatever_ expression is in those brackets) first, and additionally the analysis would somehow need to understand that this dependency only matters if `each.value.parent_key` is not null, so this seems to generalize into needing to do arbitrary expression evaluation before we've built the dependency graph, and thus we have a chicken vs. egg problem.

---

The use-case represented by _this_ issue is probably closest to the first item above: in addition to having a concurrency limit of 1, it also wants to introduce an artificial time delay between the instances.

I've always been quite loathe to introduce special cases where a particular trigger is hard-coded into OpenTofu (in this case: trigger after a certain amount of time has passed) rather than, perhaps, generalizing it to allow provider plugins to emit arbitrary signals. For example, "halt until a specific healthcheck has been reporting consistently healthy for 5 seconds" or "halt until this VM appears as a member of my Kubernetes cluster" are some obvious variations of this requirement that OpenTofu wouldn't be able to achieve without some help from a provider.

If we could convince ourselves that a purely-time-based requirement is _the main case_ (as opposed to just being a workaround for the absence of a more general feature that folks _really_ wanted, as `time_sleep` often is in practice) then we could perhaps add a time-delay-specific builtin to OpenTofu here, but we don't really have much precedent for the language runtime having such a single-purpose-focused feature and so there's definitely a non-trivial cost vs. benefit tradeoff to be made here.
